onnxruntime/contrib_ops/cpu/ngram.cc

```diff
@@ -0,0 +1,512 @@
+          ngram_size == N) {
+        if (impl_->pool_strings_.empty()) {
+          auto before_insert = impl_->int_set_.size();
+          Emplace(pool_int64s.begin() + start_idx, ngrams, ngram_size, ngram_id, impl_->int_set_);
+          ORT_ENFORCE((before_insert + ngrams) == impl_->int_set_.size(), "pool_int64s duplicate ", std::to_string(ngram_size), "-grams detected");
+        } else {
+          auto before_insert = impl_->str_set_.size();
+          Emplace(impl_->pool_strings_.begin() + start_idx, ngrams, ngram_size, ngram_id, impl_->str_set_);
+          ORT_ENFORCE((before_insert + ngrams) == impl_->str_set_.size(), "poll_strings duplicate ", std::to_string(ngram_size), "-grams detected");
+        }
+      } else {
+        ngram_id += ngrams;
+      }
+    }
+    ++ngram_size;
+  }
+}
+
+Ngram::~Ngram() {
+}
+
+void Ngram::OutputResult(OpKernelContext* ctx, const std::vector<uint32_t>& frequences) const {
+  std::vector<int64_t> output_dims;
+  output_dims.push_back(frequences.size());
+
+  TensorShape output_shape(output_dims);
+  auto Y = ctx->Output(0, output_shape);
+  auto output_data = Y->MutableData<float>();
+  const auto& w = impl_->weights_;
+  switch (impl_->mode_) {
+    case kTF: {
+      for (auto f : frequences) {
+        *output_data++ = static_cast<float>(f);
+      }
+    } break;
+    case kIDF: {
+      if (!w.empty()) {
+        assert(frequences.size() == w.size());
+        for (size_t i = 0; i < frequences.size(); ++i) {
+          *output_data++ = (frequences[i] > 0) ? w[i] : 0;
+        }
+      } else {
+        for (auto f : frequences) {
+          *output_data++ = (f > 0) ? 1.0f : 0;
+        }
+      }
+      break;
+      case kTFIDF: {
+        if (!w.empty()) {
+          assert(frequences.size() == w.size());
+          for (size_t i = 0; i < frequences.size(); ++i) {
+            *output_data++ = frequences[i] * w[i];
+          }
+        } else {
+          for (auto f : frequences) {
+            *output_data++ = static_cast<float>(f);
+          }
+        }
+      } break;
+      case kNone:  // fall-through
+      default:
+        assert(false);
+    }
+  }
+}
+
+template <typename T>
+void Ngram::ComputeImpl(OpKernelContext* ctx, size_t total_items) const {
+  const auto& impl = *impl_;
+  auto const set_end = impl.PoolEnd<T>();
+  // Frequency holder, init all to zero
+  std::vector<uint32_t> frequencies;
+  frequencies.resize(impl.output_size_, 0);
+
+  const auto N = impl.N_;
+  const auto S = impl.S_ + 1;  // Convert to distance
+  auto start_ngram_size = (impl.all_) ? impl.M_ : N;
+
+  auto X = ctx->Input<Tensor>(0);
+  auto const input_data = X->template Data<T>();
+  auto const end_data = input_data + total_items;
+  NGramItem<T> sample;
+
+  // Treat unigrams in a special way
+  if (start_ngram_size == 1) {
+    auto ngram_start = input_data;
+    while (ngram_start < end_data) {
+      sample.Clear();
+      sample.AddItem(*ngram_start);
```

Comment:
sample[0] = *ngram_start ?

Reply:
Has to be push_back as I do not resize since the size always changes and things depend on the proper size.

Reply:
Has to be push_back as I do not resize since the size always changes and things depend on the proper size.

---
In reply to: [242708266](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r242708266) [](ancestors = 242708266)